felix/dataplane/linux/bpf_ep_mgr_test.go

```diff
@@ -520,6 +538,85 @@ var _ = Describe("BPF Endpoint Manager", func() {
+			val := bpfEpMgr.polNameToRuleIDs["allowPol"]
+			Expect(val.Contains(ingRule.RuleId)).To(BeTrue())
+			Expect(val.Contains(egrRule.RuleId)).To(BeTrue())
+			Expect(bpfEpMgr.ruleIdToMatchID).To(HaveLen(2))
+			Expect(bpfEpMgr.ruleIdToMatchID[ingRule.RuleId]).To(Equal(ingRuleMatchId))
+			Expect(bpfEpMgr.ruleIdToMatchID[egrRule.RuleId]).To(Equal(egrRuleMatchId))
+			binary.LittleEndian.PutUint64(k, ingRuleMatchId)
+			binary.LittleEndian.PutUint64(v, uint64(10))
+			err := rcMap.Update(k[:], v[:])
+			Expect(err).NotTo(HaveOccurred())
+			binary.LittleEndian.PutUint64(k, egrRuleMatchId)
+			err = rcMap.Update(k[:], v[:])
+			Expect(err).NotTo(HaveOccurred())
+
+			// update the ingress rule of the policy
+			ingDenyRule := &proto.Rule{Action: "Deny", RuleId: "INGRESSDENY12345"}
+			ingDenyRuleMatchId := ruleMatchID("Ingress", "Deny", "Policy", 0, "allowPol")
+			bpfEpMgr.OnUpdate(&proto.ActivePolicyUpdate{
+				Id:     &proto.PolicyID{Tier: "default", Name: "allowPol"},
+				Policy: &proto.Policy{InboundRules: []*proto.Rule{ingDenyRule}, OutboundRules: []*proto.Rule{egrRule}},
+			})
+			Expect(bpfEpMgr.polNameToRuleIDs).To(HaveLen(1))
+			val = bpfEpMgr.polNameToRuleIDs["allowPol"]
+			Expect(val.Contains(ingDenyRule.RuleId)).To(BeTrue())
+			Expect(val.Contains(egrRule.RuleId)).To(BeTrue())
+			Expect(bpfEpMgr.dirtyRules.Contains(ingRule.RuleId)).To(BeTrue())
+			Expect(bpfEpMgr.ruleIdToMatchID).To(HaveLen(3))
+			err = bpfEpMgr.CompleteDeferredWork()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(bpfEpMgr.dirtyRules.Contains(ingRule.RuleId)).NotTo(BeTrue())
+			binary.LittleEndian.PutUint64(k, ingRuleMatchId)
+			_, err = rcMap.Get(k)
+			Expect(err).To(HaveOccurred())
+			binary.LittleEndian.PutUint64(k, egrRuleMatchId)
+			v, err = rcMap.Get(k)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(binary.LittleEndian.Uint64(v)).To(Equal(uint64(10)))
+			Expect(bpfEpMgr.ruleIdToMatchID).To(HaveLen(2))
+			Expect(bpfEpMgr.ruleIdToMatchID[ingDenyRule.RuleId]).To(Equal(ingDenyRuleMatchId))
+			Expect(bpfEpMgr.ruleIdToMatchID[egrRule.RuleId]).To(Equal(egrRuleMatchId))
+
+			// delete the policy
+			bpfEpMgr.OnUpdate(&proto.ActivePolicyRemove{Id: &proto.PolicyID{Tier: "default", Name: "allowPol"}})
+			Expect(bpfEpMgr.dirtyRules.Contains(egrRule.RuleId)).To(BeTrue())
+			Expect(bpfEpMgr.dirtyRules.Contains(ingDenyRule.RuleId)).To(BeTrue())
+			Expect(bpfEpMgr.polNameToRuleIDs).To(HaveLen(0))
+			Expect(bpfEpMgr.ruleIdToMatchID).To(HaveLen(2))
+			err = bpfEpMgr.CompleteDeferredWork()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(bpfEpMgr.ruleIdToMatchID).To(HaveLen(0))
+			binary.LittleEndian.PutUint64(k, ingRuleMatchId)
+			_, err = rcMap.Get(k)
+			Expect(err).To(HaveOccurred())
+			binary.LittleEndian.PutUint64(k, egrRuleMatchId)
+			_, err = rcMap.Get(k)
+			Expect(err).To(HaveOccurred())
+			binary.LittleEndian.PutUint64(k, ingDenyRuleMatchId)
+			_, err = rcMap.Get(k)
+			Expect(err).To(HaveOccurred())
+
+		})
+	})
+
```

Comment:
I think we should add a restart scenario with non-empty map at the start. Do we want to preserve counters across felix restarts? We definitely want to clean up and remove unused counters.